### PR TITLE
check whether ip rule exists before add

### PR DESCRIPTION
**Pull request by cadmuxe**

### CODE REVIEW

File: pkg/config/config.go [hunk 1]
```diff
@@ -100,9 +101,15 @@ func (r IPRouteConfig) Ensure(enabled bool) error {
 func (r IPRuleConfig) Ensure(enabled bool) error {
 	var err error
 	if enabled {
-		err = netlink.RuleAdd(&r.Rule)
-		if os.IsExist(err) {
-			err = nil
+		exist, e := r.exist()
```
Comment (jingax10) [inline pkg/config/config.go]: As we tested on CoS, we would run into a case to add the same rule multiple times. Can we change the logic to ensure the number of desired IP rule to be 1 when enabled is true while 0 when enabled is false?

I think the above logic is more robust as you have no assumption on what the set of IP rules looks like.

Comment (cadmuxe) [reply pkg/config/config.go]: yeah, I was debating on the same issue. will update.

Comment (cadmuxe) [reply pkg/config/config.go]: updated.

File: pkg/config/config.go [hunk 2]
```diff
@@ -100,19 +101,53 @@ func (r IPRouteConfig) Ensure(enabled bool) error {
...
+				glog.Errorf("failed to delete duplicated ip rule: $v, error: %v", r.Rule, err)
+			}
+			ruleCount--
 		}
 	}
 
 	return err
 }
 
+func (r IPRuleConfig) exist() (int, error) {
```
Comment (jingax10) [inline pkg/config/config.go]: could we have a better name other than exist, e.g., ruleCount or count?

File: pkg/config/config.go [hunk 3]
```diff
@@ -100,19 +101,53 @@ func (r IPRouteConfig) Ensure(enabled bool) error {
...
+	rules, err := netlink.RuleList(unix.AF_INET)
+	if err != nil {
+		return 0, err
+	}
+	var count int
```
Comment (jingax10) [inline pkg/config/config.go]: nit: count := 0

better to be explicitly initialized

File: pkg/config/config.go [hunk 4]
```diff
@@ -100,19 +101,53 @@ func (r IPRouteConfig) Ensure(enabled bool) error {
 func (r IPRuleConfig) Ensure(enabled bool) error {
 	var err error
 	if enabled {
-		err = netlink.RuleAdd(&r.Rule)
-		if os.IsExist(err) {
-			err = nil
+		ruleCount, e := r.exist()
```
Comment (jingax10) [inline pkg/config/config.go]: can we have a helper function, say ensureRuleCount(int), then

if enabled {
  ensureRuleCount(1)
} else {
  ensureRuleCount(0)
}

**Comment by cadmuxe** [comment]
Build failed with Go: master, succeed with 1.8, 1.9, 1.10. 
The failure due to golang

File: pkg/config/config.go [hunk 5]
```diff
@@ -98,21 +99,50 @@ func (r IPRouteConfig) Ensure(enabled bool) error {
...
+	}
+
+	for ruleCount != ensureCount {
+		if ruleCount > ensureCount {
+			if err = netlink.RuleDel(&r.Rule); err != nil {
+				glog.Errorf("failed to delete duplicated ip rule: %v, error: %v", r.Rule, err)
+			}
+			ruleCount--
+		} else {
+			if err = netlink.RuleAdd(&r.Rule); os.IsExist(err) {
```
Comment (jingax10) [inline pkg/config/config.go]: what if the error is not IsExist, can you log an err?

File: pkg/config/config.go [hunk 6]
```diff
@@ -98,21 +99,50 @@ func (r IPRouteConfig) Ensure(enabled bool) error {
...
+		return r.ensureHelper(1)
+	}
+	return r.ensureHelper(0)
+}
+
+func (r IPRuleConfig) ensureHelper(ensureCount int) error {
+	var err error
+	ruleCount, err := r.count()
+	if err != nil {
+		return err
```
Comment (jingax10) [inline pkg/config/config.go]: can you log an error?

**Comment by jingax10** [comment]
@cadmuxe  Could you please squash your commits before merge?

**Pull request merged**